fhir2dataset/graph_tools.py

```diff
@@ -40,15 +40,7 @@ def clean_euler_path(eulerian_path: list) -> list:
         list -- cleaned Eulerian path
     """  # noqa
     path = []
-    while eulerian_path:
-        edge = eulerian_path.pop(0)
-        path.append(edge)
-        list_to_pop = []
-        for ind in range(len(eulerian_path)):
-            if edge[::-1] == eulerian_path[ind] or edge == eulerian_path[ind]:
-                list_to_pop.append(ind)
-        if list_to_pop:
-            list_to_pop.sort(reverse=True)
-            for ind in list_to_pop:
-                eulerian_path.pop(ind)
+    for edge in eulerian_path:
+        if edge not in path and edge[::-1] not in path:
```

Comment:
Are edges represented by tuples/arrays?
Couldn't we use an Edge class?

Reply:
That's it! I need function `clean_euler_path` in my `join_path` function. I need this `join_path` function to know in which order I have to make my joins: one node of the graph corresponds to a table and one edge to a join that has to be made. This function allows us to create from a graph (object of class NetworkX) a list of the successive joins to be made between the table that accumulates joins and the new table to join. 

I have compared this problem to that of finding a trail in a finite graph that visits every edge exactly once (that is the definition of an eulerian path). Nevertheless, all graphs are not necessarily Eulerian graphs, so even if you eulerize a graph (which has the effect of duplicating some edges), the Eulerian path will not correspond to a single visit of all edges because edges may appear several times. And so, the `clean_euler_path` function must solve this duplication problem (I wouldn't want to make the same join twice!). By removing the duplications, we won't really have a path, but an orderly list that corresponds well to our final goal.

So, as these are "complicated" problems, I use algorithms already implemented in the NetworkX library: in the `join_path` function I do 
```
import networkx as nx

euler_graph = nx.eulerize(graph)
euler_path = list(nx.eulerian_path(euler_graph))
path = clean_euler_path(euler_path)
```
Unfortunately, the `nx.eulerian_path` function returns a tuples generator, so in this sense, they represent edges (oriented) by tuples.
What you would see is to have a structure like this? 

```
import networkx as nx
from XX import Edge

def join_path(graph: nx.Graph) -> list:
    euler_graph = nx.eulerize(graph)
    euler_path_generator = nx.eulerian_path(euler_graph)
    path = clean_euler_path(euler_path_generator)

def clean_euler_path(eulerian_path: generator) -> list:
    path = []
    for raw_edge in eulerian_path:
        if Edge(raw_edge) not in path:
            path.append(Edge(raw_edge))
    return path

```
```
class Edge(object):
    def __init__(self, tuple_edge):
        self.alias_1 = tuple_edge[0]
        self.alias_2 = tuple_edge[1]

    def __eq__(self, item):
        if isinstance(item, Edge):
            return (self.alias_1 == item.alias_1 and self.alias_2 == item.alias_2) or (
                self.alias_1 == item.alias_2 and self.alias_2 == item.alias_1
            )
        else:
            return False

    def __repr__(self):
        return f"Edge({self.alias_1},{self.alias_2})"
```

Reply:
I was thinking about something similar, yes, but if the library uses tuples, we can stick to that (it isn't too much of a mess)!